### PR TITLE
Anomaly Detection Model

### PR DESCRIPTION
**Pull request by spur19**
This Anomaly Detection model uses multivariate Gaussian distribution to predict probabilities of all records in the dataset and identify outliers. If the probability is below a certain threshold, we flag the record as an anomaly.
We use the concept of F1 score  as an evaluation metric for our model, and use it to calculate precision and recall to make our predictions more accurate. This model works well as it recognises dependencies across various features, and works particularly well if the features have a Gaussian Distribution.

### CODE REVIEW

**Comment by nitesh585** [comment]
@spur19 
It seems that most of the core function you have implemented from scratch. Instead of that you can use built-in functions because their internal implementation are more efficient and less error prone. It also helps to make your code more readable.
such as,
``` 
import numpy as np
from scipy.stats import multivariate_normal

def estimateGaussian(dataset):
    mu = np.mean(dataset, axis=0)
    sigma = np.cov(dataset.T)
    return mu, sigma

def multivariateGaussian(dataset,mu,sigma):
    p = multivariate_normal(mean=mu, cov=sigma)
    return p.pdf(dataset)
```
for calculating F1 score use `sklearn.metrics.f1_score`.
for splitting the data you may use `sklearn.model_selection.train_test_split`.

You can get 1 more :heavy_check_mark: in CI test by _Fixing style issue:_ https://intel.github.io/dffml/master/contributing/style.html :sweat_smile:

**Comment by spur19** [comment]
@nitesh585 
Yeah, seems like I missed a trick there. Thanks for the help :)

File: model/scratch/dffml-model-anomalydetection/tests/test1.csv [hunk 1]
```diff
@@ -0,0 +1,92 @@
+A,B,C,D,E,F,G,H,I,J,K,Y
```
Comment (pdxjohnny) [inline model/scratch/dffml-model-anomalydetection/tests/test1.csv:1]: Let's have the test and train data generated by some Python function. Let cut it down to 2 features and then Y.

File: model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    Features,
+    ModelNotTrained,
+    Record,
+    SimpleModel,
+    SourcesContext,
+    config,
+    entrypoint,
+    field,
+)
+
```
Comment (pdxjohnny) [inline model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py:19]: Let's make sure we find out if there is any source and it's licensing we might want to attribute some of this code / ideas being from

File: model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        F1val, epsilon = selectThreshold(Yval, pval)
+
+        outliers = p < epsilon
+
+        listOfOl = findIndices(outliers)  # outliers in training set
+        # print(F1val)
+        # Save listOfOl and F1 score
+        self.storage["anomalies"] = (
+            listOfOl,
```
Comment (pdxjohnny) [inline model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py:160]: Looks like this never gets used

```suggestion
```

File: model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+            record_data = []
+            for feature in record.features(self.features).values():
+                record_data.extend([feature] if np.isscalar(feature) else feature)
+
+            X.append(record_data)
+            Y.append(record.feature(self.parent.config.predict.name))
+
+        self.logger.debug("Number of test records: %d", len(X))
+
+        nof = len(self.features)  # number of features
```
Comment (pdxjohnny) [inline model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py:192]:
```suggestion
        # Number of features
        nof = len(self.features)
```

File: model/scratch/dffml-model-anomalydetection/LICENSE [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Copyright (c) 2019 spur
```
Comment (0dust) [inline model/scratch/dffml-model-anomalydetection/LICENSE:1]:
```suggestion
Copyright (c) 2019 Intel, spur
```

File: model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    try:
+        fp = np.sum(((pred == 1) & (Yval == 0)).astype(int))
+        tp = np.sum(((pred == 1) & (Yval == 1)).astype(int))
+        fn = np.sum(((pred == 0) & (Yval == 1)).astype(int))
+
+        prec = tp / (tp + fp)
+        rec = tp / (tp + fn)
+        F1 = 2 * prec * rec / (prec + rec)
+    except ZeroDivisionError:
+        print("Warning : dividing by zero!")
```
Comment (0dust) [inline model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py:97]: You may want to raise the error here instead of printing.

File: model/scratch/dffml-model-anomalydetection/setup.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+# Boilerplate to load commonalities
+spec = importlib.util.spec_from_file_location(
+    "setup_common", os.path.join(os.path.dirname(__file__), "setup_common.py")
+)
+common = importlib.util.module_from_spec(spec)
+spec.loader.exec_module(common)
+
+common.KWARGS["entry_points"] = {
+    "dffml.model": [f"myslr = {common.IMPORT_NAME}.myslr:MySLRModel"]
```
Comment (0dust) [inline model/scratch/dffml-model-anomalydetection/setup.py:13]: Register "anomalydetection" entrypoint. You can check https://github.com/intel/dffml/blob/0b84ca2af45671a0ca3ffce8485a1a2f10b1a7c5/model/scikit/setup.py#L83 for reference.

**Comment by pdxjohnny** [comment]
- Here we did `dffml service dev model create`. However, we already have the model/scratch package. Therefore, we should take the implementation
  - model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py
- And move it to
  - model/scratch/dffml_model_scratch/anomalydetection.py

File: model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+@config
+class AnomalyModelConfig:
+    features: Features = field("Features to train on")
+    predict: Feature = field("Label or the value to be predicted")
+    directory: pathlib.Path = field("Directory where state should be saved")
+    k: float = field("Validation set size", default=0.8)
+
+
+@entrypoint("anomalydetection")
+class AnomalyModel(SimpleModel):
```
Comment (pdxjohnny) [inline model/scratch/dffml-model-anomalydetection/dffml_model_anomalydetection/anomalydetection.py:111]: https://intel.github.io/dffml/master/contributing/consoletest.html

For writing out a file:


```rst
.. code-block::
    :test:
    :filepath: training.csv

    x,y
    0,1
```

**Comment by pdxjohnny** [comment]
I can't remember what was left to do here, can you remind me?

**Comment by spur19** [comment]
> I can't remember what was left to do here, can you remind me? Or are we ready to merge, if so I'll give it another review.

Nope, we resolved the `consoletest` issue on Tuesday, so I guess we're done  :)

**Comment by pdxjohnny** [comment]
@spur19 Can you merge in master? Thanks!

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ### Added
+- New model for Anomaly Detection
```
Comment (pdxjohnny) [inline CHANGELOG.md:9]:
```suggestion
- New model for Anomaly Detection
```

**Comment by pdxjohnny** [comment]
Great work! Thank you!
